File: algorithms/[36]有效的数独/solution.ts
function isValidSudoku(board: string[][]): boolean {
  let ch: string;

  for (let i = 0; i < 9; i += 1) {
    const rowNums: Set<string> = new Set();
    const colNums: Set<string> = new Set();
    const cubeNums: Set<string> = new Set();

    for (let j = 0; j < 9; j += 1) {
      ch = board[i][j];
      if (ch !== '.') {
        if (rowNums.has(ch)) return false;
        rowNums.add(ch);
      }

      ch = board[j][i];
      if (ch !== '.') {
        if (colNums.has(ch)) return false;
        colNums.add(ch);
      }

      let row = Math.floor(i / 3) * 3 + Math.floor(j / 3);
      let col = (i % 3) * 3 + (j % 3);
      ch = board[row][col];
      if (ch !== '.') {
        if (cubeNums.has(ch)) return false;
        cubeNums.add(ch);
      }
    }
  }

  return true;
}

File: algorithms/[36]有效的数独/solution.py
# topics = ["哈希表"]

from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for i in range(9):
            rowNums = set()
            colNums = set()
            cubeNums = set()

            for j in range(9):
                ch = board[i][j]
                if ch != '.':
                    if ch in rowNums:
                        return False
                    rowNums.add(ch)

                ch = board[j][i]
                if ch != '.':
                    if ch in colNums:
                        return False
                    colNums.add(ch)

                row = i // 3 * 3 + j // 3
                col = i % 3 * 3 + j % 3
                ch = board[row][col]
                if ch != '.':
                    if ch in cubeNums:
                        return False
                    cubeNums.add(ch)

        return True
